sigstore/_verify/policy.py

```diff
@@ -0,0 +1,233 @@
+
+    An empty list of child policies is considered trivially invalid.
+    """
+
+    def __init__(self, children: list[VerificationPolicy]):
+        self._children = children
+
+    def verify(self, cert: Certificate) -> VerificationResult:
+        verified = any(child.verify(cert) for child in self._children)
+        if verified:
+            return VerificationSuccess()
+        else:
+            return VerificationFailure(
+                reason=f"0 of {len(self._children)} policies succeeded"
+            )
+
+
+class AllOf:
+    """
+    The "all of" policy, corresponding to a logical AND between child
+    policies.
+
+    An empty list of child policies is considered trivially invalid.
+    """
+
+    def __init__(self, children: list[VerificationPolicy]):
+        self._children = children
+
+    def verify(self, cert: Certificate) -> VerificationResult:
+        # Without this, we'd consider empty lists of child policies trivially valid.
+        # This is almost certainly not what the user wants and is a potential
+        # source of API misuse, so we explicitly disallow it.
+        if len(self._children) < 1:
+            return VerificationFailure(reason="no child policies to verify")
+
+        # NOTE(ww): We need the cast here because MyPy can't tell that
+        # `VerificationResult.__bool__` is invariant with
+        # `VerificationSuccess | VerificationFailure`.
+        results = [child.verify(cert) for child in self._children]
+        failures = [
+            cast(VerificationFailure, result).reason for result in results if not result
+        ]
+        if len(failures) > 0:
+            inner_reasons = ", ".join(failures)
+            return VerificationFailure(
+                reason=f"{len(failures)} of {len(self._children)} policies failed: {inner_reasons}"
+            )
+        return VerificationSuccess()
+
+
+class Identity:
+    """
+    Verifies the certificate's "identity", corresponding to the X.509v3 SAN.
+    Identities are verified modulo an OIDC issuer, so the issuer's URI
+    is also required.
+
+    Supported SAN types include emails, URIs, and Sigstore-specific "other names".
+    """
+
+    def __init__(self, *, identity: str, issuer: str):
+        self._identity = identity
+        self._issuer = OIDCIssuer(issuer)
+
+    def verify(self, cert: Certificate) -> VerificationResult:
+        issuer_verified: VerificationResult = self._issuer.verify(cert)
+        if not issuer_verified:
+            return issuer_verified
+
+        san_ext = cert.extensions.get_extension_for_class(SubjectAlternativeName)
+        verified = (
+            self._identity in san_ext.value.get_values_for_type(RFC822Name)
+            or self._identity
+            in san_ext.value.get_values_for_type(UniformResourceIdentifier)
+            or OtherName(_OTHERNAME_OID, self._identity.encode())
+            in san_ext.value.get_values_for_type(OtherName)
```

Comment:
Yeah, this was a concern I had originally -- I can't find the full conversation now, but @znewman01 and I discussed this a while back.

I _think_ this will be okay, since each identifier's format is unique to its type: emails are always in RFC822 form, usernames are in `ident!service` form, and URIs are always in URI form. We rely on Fulcio enforcing these rules, per https://github.com/sigstore/fulcio/pull/802.

Reply:
Yeah, here's the original RFC doc: https://docs.google.com/document/d/1o8_bXIygufgiohJGlmBzqF4_BnXCTfgh4ILgJFJxYRs/edit?resourcekey=0-YEar3v67uoT31kj83dCVvA#

Reply:
(xref https://github.com/sigstore/fulcio/issues/716)